src/actions/actions.ts

```diff
@@ -590,10 +590,17 @@ class CommandInsertInInsertMode extends BaseCommand {
         vimState.cursorStartPosition = leftPosition;
       }
     } else {
-      await TextEditor.insert(char, vimState.cursorPosition);
+      const selection = vscode.window.activeTextEditor.selection;
+      if (selection.isEmpty) {
+        await TextEditor.insert(char, vimState.cursorPosition);
+      } else {
+        await TextEditor.replace(selection, char);
+      }
+
+      const cursorPosition = vscode.window.activeTextEditor.selection.end;
 
-      vimState.cursorStartPosition = Position.FromVSCodePosition(vscode.window.activeTextEditor.selection.start);
-      vimState.cursorPosition = Position.FromVSCodePosition(vscode.window.activeTextEditor.selection.start);
+      vimState.cursorStartPosition = Position.FromVSCodePosition(cursorPosition);
```

Comment:
But why would you even need to convert cursorPostition from a vscode position in the first place - it should already be one.